packages/go_router/lib/src/route.dart

```diff
@@ -97,10 +97,14 @@ import 'typedefs.dart';
 abstract class RouteBase {
   const RouteBase._({
     this.routes = const <RouteBase>[],
+    this.onExit,
   });
 
   /// The list of child routes associated with this route.
   final List<RouteBase> routes;
+
+  /// Return false to prevent the route from being popped.
+  final Future<bool> Function(BuildContext context)? onExit;
```

Comment:
This can be FutureOr so that it can either be sync or async

Reply:
also should we allow people to do more than just pop or not? I was curious whether we need to support cases like redirect to a different route or else